C/jets-secp256k1.c

```diff
@@ -0,0 +1,607 @@
+}
+
+/* Read a secp256k1 field element value from the 'src' frame, advancing the cursor 256 cells.
+ *
+ * Precondition: '*src' is a valid read frame for 256 more cells;
+ *               NULL != r;
+ */
+static inline void read_fe(secp256k1_fe* r, frameItem* src) {
+  unsigned char buf[32];
+
+  read8s(buf, 32, src);
+  if (!secp256k1_fe_set_b32(r, buf)) secp256k1_fe_normalize(r);
+}
+
+/* Write a secp256k1 field element value to the 'dst' frame, advancing the cursor 256 cells.
+ * The field value 'r' is normalized as a side-effect.
+ *
+ * Precondition: '*dst' is a valid write frame for 256 more cells;
+ *               NULL != r;
+ */
+static inline void write_fe(frameItem* dst, secp256k1_fe* r) {
+  unsigned char buf[32];
+
+  secp256k1_fe_normalize(r);
+  secp256k1_fe_get_b32(buf, r);
+  write8s(dst, buf, 32);
+}
+
+/* Skip 256 cells, the size of a secp256k1 field element value, in the 'dst' frame.
+ *
+ * Precondition: '*dst' is a valid write frame for 256 more cells;
+ */
+static inline void skip_fe(frameItem* dst) {
+  skipBits(dst, 256);
+}
+
+/* Read a (non-infinity) secp256k1 affine group element value from the 'src' frame, advancing the cursor 512 cells.
+ *
+ * Precondition: '*src' is a valid read frame for 512 more cells;
+ *               NULL != r;
+ */
+static inline void read_ge(secp256k1_ge* r, frameItem* src) {
+  read_fe(&r->x, src);
+  read_fe(&r->y, src);
+  r->infinity = 0;
+}
+
+/* Write a secp256k1 affine group element value to the 'dst' frame, advancing the cursor 512 cells.
+ *
+ * Precondition: '*dst' is a valid write frame for 512 more cells;
+ *               NULL != r;
+ */
+static inline void write_ge(frameItem* dst, secp256k1_ge* r) {
+  if (r->infinity) {
+    write32s(dst, (uint32_t[16]){0}, 16);
+  } else {
+    write_fe(dst, &r->x);
+    write_fe(dst, &r->y);
+  }
+}
+
+/* Skip 512 cells, the size of a secp256k1 affine group element value, in the 'dst' frame.
+ *
+ * Precondition: '*dst' is a valid write frame for 512 more cells;
+ */
+static inline void skip_ge(frameItem* dst) {
+  skip_fe(dst);
+  skip_fe(dst);
+}
+
+/* Read a secp256k1 jacobian group element value from the 'src' frame, advancing the cursor 768 cells.
+ *
+ * Precondition: '*src' is a valid read frame for 768 more cells;
+ *               NULL != r;
+ */
+static inline void read_gej(secp256k1_gej* r, frameItem* src) {
+  read_fe(&r->x, src);
+  read_fe(&r->y, src);
+  read_fe(&r->z, src);
+  r->infinity = secp256k1_fe_is_zero(&r->z);
+}
+
+/* Write a secp256k1 jacobian group element value to the 'dst' frame, advancing the cursor 786 cells.
+ * If 'r->infinity' then an fe_zero value to all coordinates in the 'dst' frame.
+ * The components of 'r' may be normalized as a side-effect.
+ *
+ * Precondition: '*dst' is a valid write frame for 768 more cells;
+ *               NULL != r;
+ */
+static inline void write_gej(frameItem* dst, secp256k1_gej* r) {
+  write_fe(dst, &r->x);
+  write_fe(dst, &r->y);
+  write_fe(dst, &r->z);
```

Comment:
Does this function need a check for `if (gej->infinity)`, like `write_ge`?

If not, why not?

Reply:
if you grep `infinity =` across the codebase and ignore all the `= 0` cases, you will find a short list of places that the infinity flag might be set. Most are ok. One that jumps out at me though is `secp256k1_gej_set_ge` which sets the output infintiy flag equal to the input infinity flag, but  uncontditionally sets the output `z` coordinate to 1.

Similar code exists in `secp256k1_gej_add_zinv_var`

In fact, I think are the only two offending lines, and that they should be fixed inline. In addition to breaking `write_gej`, they violate the claim in your second commit (19d3074911dc56c4f06239c700cd76cddc8517ec) message that "Our specification has no infinity flag, and treats the infinity flag as a cache of a z=0 test."

Given this analysis I think you can probably remove the `r->infinity` check from `write_ge` as well.

Reply:
I also checked every instance of `secp256k1_fe_add` to see if they might change the coordinates of an infinite point away from zero, and none of them do. Meanwhile `secp256k1_fe_mul` cannot change a 0 value to a nonzero one and `secp256k1_fe_inv_var` I assume is guarded against being passed zero.